crates/compiler/src/obj_gen/mod.rs

```diff
@@ -2545,73 +2781,138 @@ impl ObjectGenerator {
             .collect::<StdResult<Vec<_>>>()?;
-        let return_type = LLVMType::try_from(instruction.get_type())?;
+        let ret_type = LLVMType::try_from(instruction.get_type())?;
+        let func_type = LLVMFunction::new(ret_type.clone(), arg_types.as_slice());
 
-        // If it is NOT one of these, we have to actually generate the call, and to do
-        // that we need to know whether it is an internal or external call.
-        let function_ref = if let Some(id) = func_ctx.module_functions().get(&function_name) {
-            BlockRef::Local(*id)
-        } else if let Some(polyfill_name) =
-            self.polyfills
-                .get_polyfill(&function_name, &operand_types, &return_type)
-        {
-            BlockRef::Builtin(polyfill_name.to_string())
+        // We also need our arguments and return values, remembering to register the
+        // latter in the context.
+        let mut function_inputs = function_arguments
+            .iter()
+            .map(|op| util::get_var_or_const(op, bb, func_ctx))
+            .collect::<Result<Vec<_>>>()?;
+        let output_type_flo = ObjectContext::flo_type_of(&ret_type)?;
+        let output_var = bb.add_variable(output_type_flo.clone());
+
+        // The name for the output variable may not exist depending on where this call
+        // comes from, so we may have to create one.
+        let output_name = if let Some(name) = instruction.get_name() {
+            name.to_str()?.to_string()
         } else {
-            BlockRef::External(function_name)
+            self.name_supply.borrow_mut().allocate()
         };
+        func_ctx.register_local(output_var, &output_name, output_type_flo);
 
-        // We then need to put together the operands by grabbing the variables from the
-        // context and passing them in. Given that the function name is the LAST
-        // operand, it is the one we want to ignore.
-        let inputs = value_operands
-            .into_iter()
-            .map(|op| util::get_var_or_const(&op, bb, func_ctx))
-            .collect::<Result<Vec<_>>>()?;
+        let function_outputs =
+            vec![bb.add_variable(ObjectContext::flo_type_of(&ret_type.clone())?)];
 
-        // We also the need to handle providing the return value, if one exists, using
-        // the type and whether the instruction has a name telling us.
-        let outputs = util::make_opcode_output(&instruction, bb, func_ctx)?;
+        // To call the function we need to resolve the block reference through which it
+        // will be called.
+        let func_ref = if let Some(id) = func_ctx.module_functions().get(&func_name) {
+            // In this case, it is a function in the same module that we can issue a direct
+            // call to.
+            BlockRef::Local(*id)
+        } else if let Some(name) = self.polyfills.get_polyfill(&func_name, &arg_types, &ret_type) {
+            // In this case, it is a call to a known polyfill in the runtime.
```

Comment:
Are we intending to let user-code call our polyfills as compiler-builtins? Most compilers allow overriding of _named_ builtins via user code; in which case this branch would also need to check if this is a defined-name or not.

Reply:
User code should be able to call them, yes. We'll need that for the chain support Rust library later. I was not going to allow overriding of these however, so the intent of this is correct. We should probably do an on-ingest check against names, but I will leave that for later. Filed as #135.